### PR TITLE
Adding a cli transport option for the bigip_command module.

### PR DESCRIPTION
**Pull request by jmighion**
##### SUMMARY

Adds the ability to run commands via the CLI on F5 devices through the bigip_command module.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

bigip_command
##### ANSIBLE VERSION

```
 ansible --version
ansible 2.5.0 (f5/bigip_command_cli 31c42d9dde) last updated 2017/09/14 15:17:11 (GMT -700)
  config file = None
  configured module search path = [u'/Users/james/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/james/Documents/git/ansible/lib/ansible
  executable location = /Users/james/Documents/git/ansible/bin/ansible
  python version = 2.7.13 (default, Jul 18 2017, 09:17:00) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

### CODE REVIEW

**Comment by jmighion** [comment]
Shippable unstable. Closing and reopening to restart.

File: lib/ansible/plugins/action/bigip.py
Comment (ganeshrn) [inline lib/ansible/plugins/action/bigip.py]: In 2.5 the logic to load provider arguments is moved at a common place in [module_utils/network_common.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios.py#L48)
Please refer [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios.py#L48).

File: lib/ansible/modules/network/f5/bigip_command.py
Comment (ganeshrn) [inline lib/ansible/modules/network/f5/bigip_command.py:67]: Is it better to add `transport` as part of common parameters?

**Review by ganeshrn — CHANGES_REQUESTED** [review]
Reuse load_provider from `module_utils/network_common.py` and other minor change.

File: lib/ansible/modules/network/f5/bigip_command.py
Comment (gundalow) [reply lib/ansible/modules/network/f5/bigip_command.py:67]: That's a good suggestion, though only if `transport` is truly supported by all F5 modules

File: lib/ansible/modules/network/f5/bigip_command.py
Comment (gundalow) [inline lib/ansible/modules/network/f5/bigip_command.py:74]: @caphrim007 is `rest` how you describe the interface?

File: lib/ansible/plugins/action/bigip.py
Comment (gundalow) [inline lib/ansible/plugins/action/bigip.py]: Is this going to cause existing 2.4 playbooks to fail?

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
Tests will need adding for this before we can merge it.

As we don't currently have the ability to run F5 tests ourselves, the tests will need adding to F5's repo.

I believe the correct repo is https://github.com/F5Networks/f5-ansible/tree/devel/test/integration

Please check with @caphrim007 for how this should be done

File: lib/ansible/modules/network/f5/bigip_command.py
Comment (jmighion) [reply lib/ansible/modules/network/f5/bigip_command.py:67]: I was specifically told to move it to the module here to avoid breaking other F5 modules.

**Comment by jmighion** [comment]
This has already been PR'd and approved by @caphrim007 in https://github.com/F5Networks/f5-ansible/pull/330
~~I'll get to work on the load provider from network_common.py today.~~ Won't work with any modules since the provider is added to the task args.

File: lib/ansible/plugins/action/bigip.py
Comment (jmighion) [reply lib/ansible/plugins/action/bigip.py]: I'll have to defer to @caphrim007 on that one. My test environment doesn't allow me to validate the rest api, but I was originally told by @caphrim007 that the existing modules in 2.4 were not affected.

File: lib/ansible/modules/network/f5/bigip_command.py
Comment (caphrim007) [reply lib/ansible/modules/network/f5/bigip_command.py:67]: yes, do **not** add it to all modules.

File: lib/ansible/modules/network/f5/bigip_command.py
Comment (caphrim007) [reply lib/ansible/modules/network/f5/bigip_command.py:74]: @gundalow yes

File: lib/ansible/plugins/action/bigip.py
Comment (caphrim007) [reply lib/ansible/plugins/action/bigip.py]: @gundalow ahh, true.

F5 modules are not only used with connection local. They are also used with connection ssh and delegate_to. I think the above would fail if one used delegation.

File: lib/ansible/plugins/action/bigip.py
Comment (jmighion) [reply lib/ansible/plugins/action/bigip.py]: I don't believe we can do that since the provider isn't in the default args. The one in network_common adds the provider `args['provider'] = provider` which will result in the other F5 modules erroring with `Unsupported parameters for (bigip_config) module: provider`

Comment (ganeshrn) [reply lib/ansible/plugins/action/bigip.py]: I see the issue, in this case, `F5_COMMON_ARGS` is similar to `provider`  but not required to be updated in task args. For other platforms `provider` need to be part of task args as provider spec has some default values which are directly referenced in the module code. 

One way to avoid duplicating this code is to delete `provider` from task args in `bigip` action plugin before calling `run()`.

If this is applicable for other platforms in future I think it is better to move the line that assigns `provider` to task args in each of the action plugins. Thoughts?

Comment (jmighion) [reply lib/ansible/plugins/action/bigip.py]: I'll let @caphrim007 decide since he's the one maintaining all the F5 modules.
I'm ok with deleting `provider` from the task args in the `bigip` action plugin. Always like reusing code vs duplicating it.

Comment (caphrim007) [reply lib/ansible/plugins/action/bigip.py]: deleting and re-using the existing code is fine.

**Review by ganeshrn — APPROVED** [review]
Looks good

**Comment by jmighion** [comment]
@gundalow Testing and validation has passed according to @caphrim007. Let me know if there's anything else you need to approve this.

**Comment by gundalow** [comment]
Unit tests will need updating to defend this functionality.

What what is currently tested, please see
https://codecov.io/gh/ansible/ansible/src/devel/lib/ansible/modules/network/f5/bigip_command.py
https://codecov.io/gh/ansible/ansible/src/devel/lib/ansible/module_utils/f5_utils.py

**Comment by caphrim007** [comment]
fyi, I gave a "lgtm" in the f5-ansible module where the initial tests were added.

https://github.com/F5Networks/f5-ansible/pull/337

**Comment by jmighion** [comment]
@gundalow Let me know if you need more

**Comment by jmighion** [comment]
@gundalow Let me know if you need more

**Pull request merged**
